Add "unsafe" AsyncRequestBody constructors for better performance when concurrent modification isn't a worry

StephenFlavin opened the pull request:
The static methods in `AsyncRequestBody` do some defensive copying and transforming of types that hamper performance and utilise more memory than needed if the user has already ensured either that concurrent modification won't happen or that it doesn't matter, this pr addresses that with new `AsyncRequestBody` constructors.

This also fixes an unnecessary copy being made in `AsyncRequestBody#fromByteBuffer` (https://github.com/aws/aws-sdk-java-v2/issues/1735) and `AsyncRequestBody#fromString` as a side effect.

## Motivation and Context

Memory consumption and performance.

Related Issues:
 - https://github.com/aws/aws-sdk-java-v2/issues/3193
 - https://github.com/aws/aws-sdk-java-v2/issues/1735
 - https://github.com/aws/aws-sdk-java-v2/issues/3928


## Modifications

Adds `AsyncRequestBody#fromBytesUnsafe(byte[])`, `AsyncRequestBody#fromByteBufferUnsafe(ByteBuffer)` and `AsyncRequestBody#fromByteBuffersUnsafe(ByteBuffer...)` as no copy alternatives to the existing methods.

Adds `AsyncRequestBody#fromByteBuffers(ByteBuffer...)`, `AsyncRequestBody#fromRemainingByteBuffer`, `AsyncRequestBody#fromRemainingByteBufferUnsafe`, `AsyncRequestBody#fromRemainingByteBuffers` and `AsyncRequestBody#fromRemainingByteBuffersUnsafe` methods as implementations that respect the current buffer positions.

Updates `AsyncRequestBodyTest` to junit 5.

Removes `ByteArrayAsyncRequestBody` 

Adds `ByteBuffersAsyncRequestBody` with static builders for `ByteBuffer`, `ByteBuffer[]` and `byte[]`.
 - calls to `ByteBuffersAsyncRequestBody#subscribe` are repeatable (not sure if this is a requirement but would be a breaking change vs `ByteArrayAsyncRequestBody`)
 - `ByteBuffersAsyncRequestBody#subscribe` creates a thread safe subscription.
 - `ByteBuffer[]` allows a single file > `Integer.MAX_VALUE` to be transferred.
 - A benefit of the implementation is `ByteBuffersAsyncRequestBody` doesn't copy an off heap ByteBuffer until it's requested by the subscription so depending on the subscriber less data will be on heap at a time.
     -   _I would like to know why off heap buffers need to be copied at all (see: issue https://github.com/aws/aws-sdk-java-v2/issues/3928)_

Adds `BinaryUtils#immutableCopyOf`, `BinaryUtils#immutableCopyOfRemaining` and `BinaryUtils#toNonDirectBuffer` convenience methods.

## Testing
Unit tests

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist
- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] ~Local run of `mvn install` succeeds~ https://github.com/aws/aws-sdk-java-v2/issues/3926
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new ~and existing~ tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License
- [x] I confirm that this pull request can be released under the Apache 2 license

millems commented on core/sdk-core/src/main/java/software/amazon/awssdk/core/async/AsyncRequestBody.java:
```diff
@@ -143,25 +143,33 @@ static AsyncRequestBody fromString(String string) {
      * @return AsyncRequestBody instance.
      */
     static AsyncRequestBody fromBytes(byte[] bytes) {
-        return new ByteArrayAsyncRequestBody(bytes, Mimetype.MIMETYPE_OCTET_STREAM);
+        return ByteBuffersAsyncRequestBody.from(bytes);
     }
```
Similar to `SdkBytes`, can we keep the defensive copy here, but introduce a `fromBytesUnsafe` to create it without the defensive copy, so it's more obvious to the caller that they should not mutate the byte array after it is passed?

millems commented on core/sdk-core/src/main/java/software/amazon/awssdk/core/async/AsyncRequestBody.java:
```diff
@@ -143,25 +143,33 @@ static AsyncRequestBody fromString(String string) {
      * @return AsyncRequestBody instance.
      */
     static AsyncRequestBody fromByteBuffer(ByteBuffer byteBuffer) {
-        return fromBytes(BinaryUtils.copyAllBytesFrom(byteBuffer));
+        return ByteBuffersAsyncRequestBody.of(null, byteBuffer);
+    }
```
Same here. Can we add `fromByteBufferUnsafe` instead of changing the default behavior?

StephenFlavin replied:
yes these can all be renamed.
I thought it might be an issue to change the contracts here, although, I'm not sure if `Unsafe` is the right word in this context, as `DirectByteBuffers` are used with `sun.misc.Unsafe` to allow de-allocation of their off heap memory and it may lead to confusion.
I am finding it difficult to come up with another name though, `fromByteBufferDirect` (same issue), `fromMutableByteBuffer` maybe 🤷

One thing to note is that `AsyncRequestBody#fromFile` doesn't provide the same guarantee regarding concurrent modification.

What are your thoughts on the defensive copying?
personally I would say the developer can apply defensive copies themselves before calling these methods as it's not a complex operation, and I'm wondering if this change could be rolled into a major version bump with other changes such as what I've outlined on https://github.com/aws/aws-sdk-java-v2/issues/3928

millems replied:
>  I thought it might be an issue to change the contracts here, although, I'm not sure if Unsafe is the right word in this context,DirectByteBuffers are used with sun.misc.Unsafe to allow de-allocation of their off heap memory and it may lead to confusion.

Yeah, "Unsafe" is meant to illustrate a clear delineation between "the SDK is making no assumptions about your input" and "the SDK is making some critical assumptions about your input". It doesn't have anything to do with `sun.misc.Unsafe`, although I agree that might be one potential confusion. Since we've already committed to the "unsafe" naming elsewhere for this case, though, I'd favor sticking with it.

> One thing to note is that AsyncRequestBody#fromFile doesn't provide the same guarantee regarding concurrent modification.

That's a good point. It is indeed inconsistent. I suppose doing things from disk and doing things from memory were just two different states of mind during our development.

> What are your thoughts on the defensive copying? personally I would say the developer can apply defensive copies themselves before calling these methods as it's not a complex operation, and I'm wondering if this change could be rolled into a major version bump with other changes such as what I've outlined on https://github.com/aws/aws-sdk-java-v2/issues/3928

I think it depends on the audience. If we can assume the developer is thoughtful and experienced, then I agree it's most intuitive to not do a defensive copy. Most people we talk with here on github are the thoughtful and experienced types.

Our assumption about the AWS SDK's customers is that they come from all levels of in-a-hurry and experience. A recently-hired intern, who has never worked with Java, and needs to finish up some work that's leftover from the person that they replaced probably won't be thinking about the mutability of the data they're passing around. They're struggling getting through all of the other AWS complexities that exist. We don't want them to miss their deadlines fighting a data mutability bug and then say "that's dumb, why does Java/AWS/the computer/etc. allow that to happen?" There's already enough places for customer's code to go wrong.

Our solution has been to make the obvious path "safe" and the less-obvious path trade some of that safety for performance. We make them use the method called "unsafe" if they want to prevent the copying.

StephenFlavin replied:
Thank you for the great response, I'll make the changes you suggest as soon as I get a free moment

StephenFlavin replied:
done

millems commented on core/sdk-core/src/main/java/software/amazon/awssdk/core/async/AsyncRequestBody.java:
```diff
@@ -143,25 +143,33 @@ static AsyncRequestBody fromString(String string) {
+     * @param byteBuffers ByteBuffer[] to send to the service.
+     * @return AsyncRequestBody instance.
+     */
+    static AsyncRequestBody fromByteBuffers(ByteBuffer... byteBuffers) {
```
Can we call this `fromByteBuffersUnsafe`?

A non-blocking observation: This method makes an interesting point, that we don't have a way of joining AsyncRequestBody implementations. This makes it work for byte buffers, but strings, byte arrays, etc. still need to be concatenated.

I almost wonder if we should have a `AsyncRequestBody append(AsyncRequestBody)` method on `AsyncRequestBody` that appends the data from one body to another, so that we don't have to have special treatment of byte buffers here.

`AsyncRequestBody.fromByteBuffer(firstBuffer).append(AsyncRequestBody.fromByteBuffer(secondBuffer));`

I don't think that would obviate the need for a simple way of joining multiple, like we have here, though. If you had 10 byte buffers you wanted to join, that seems like a pain.

Again, I don't expect you to change anything here for this PR... just documenting an observation.

StephenFlavin replied:
fyi, I added both the safe and unsafe implementation for this in the PR

millems commented on core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/ByteArrayAsyncRequestBody.java:
```diff
@@ -30,8 +30,11 @@
  */
 @SdkInternalApi
+@Deprecated
 public final class ByteArrayAsyncRequestBody implements AsyncRequestBody {
```
If we no longer use this at all, we can delete it. Internal APIs don't need to be left for backwards-compatibility.

millems commented on core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/ByteBuffersAsyncRequestBody.java:
```diff
@@ -0,0 +1,136 @@
+
+                            do {
+                                ByteBuffer buffer = buffers[i];
+                                if (!buffer.hasArray()) {
+                                    buffer = ByteBuffer.wrap(BinaryUtils.copyBytesFrom(buffer));
+                                }
```
Is this because we have assumptions downstream of this publisher that the buffer `hasArray`?

StephenFlavin replied:
I've done this to maintain how the existing code will convert direct buffers (off heap) to non-direct (on heap) buffers see: https://github.com/aws/aws-sdk-java-v2/issues/3928
I've tested with the Netty and CRT clients and neither seem to have any issues handling off-heap buffers but I was attempting to separate out the zero-copy functionality to future PR's

millems replied:
> I've done this to maintain how the existing code will convert direct buffers (off heap) to non-direct (on heap) buffers see: https://github.com/aws/aws-sdk-java-v2/issues/3928
> I've tested with the Netty and CRT clients and neither seem to have any issues handling off-heap buffers but I was attempting to separate out the zero-copy functionality to future PR's

Got it, I agree it makes sense to keep it separate for now.

millems commented on core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/ByteBuffersAsyncRequestBody.java:
```diff
@@ -0,0 +1,136 @@
+                                s.onComplete();
+                            }
+                        } else {
+                            s.onError(new IllegalArgumentException("§3.9: non-positive requests are not allowed!"));
```
`completed.set(true)` to drop future `request` calls?

StephenFlavin replied:
In hindsight I'm not sure if this would be correct, I lifted this logic from `ByteArrayAsyncRequestBody` so there's no logic change here, and I'm not sure if the this subscription would have to be marked as completed if it errors.
What do you think?

millems replied:
I think it's correct, but this case probably never happens so I doubt we're testing it. If it's already like this, I'm not really opposed to leaving it like this. We need to move this to an easier-to-use publisher pattern anyway.

millems commented on core/sdk-core/src/main/java/software/amazon/awssdk/core/async/AsyncRequestBody.java:
```diff
@@ -150,18 +151,69 @@ static AsyncRequestBody fromString(String string) {
+    }
+
+    /**
+     * Creates a {@link AsyncRequestBody} from a {@link ByteBuffer}.
```
Nit: Can we add similar wording to what we have in [SdkBytes](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/core/SdkBytes.html)'s unsafe methods? Clarifying why it is called "unsafe", for example.

StephenFlavin replied:
Done, I've updated each methods javadoc to be consistent with this wording

millems commented on utils/src/main/java/software/amazon/awssdk/utils/BinaryUtils.java:
```diff
@@ -117,6 +117,60 @@ public static ByteArrayInputStream toStream(ByteBuffer byteBuffer) {
+     * The new buffer's position will be set to the position of the given {@code ByteBuffer}, but the mark if defined will be
+     * ignored.
+     * <p>
+     * Note: this method intentionally converts direct buffers to non-direct, see
+     * <a href="https://github.com/aws/aws-sdk-java-v2/issues/3928">issue #3928</a>
```
Nit: I don't know if we should bake this intention into this method, since it will likely be used elsewhere in the future. Can we just say that this is the current behavior, and leave it at that?

StephenFlavin replied:
should I add a note that this may change in future?

StephenFlavin replied:
done

millems commented on core/sdk-core/src/main/java/software/amazon/awssdk/core/async/AsyncRequestBody.java:
```diff
@@ -143,29 +144,96 @@ static AsyncRequestBody fromString(String string) {
+    }
+
+    /**
+     * Creates an {@link AsyncRequestBody} from a {@link ByteBuffer}. This will copy the contents of the {@link ByteBuffer} to
+     * prevent modifications to the provided {@link ByteBuffer} from being reflected in the {@link AsyncRequestBody}.
+     * <p> The position is set to 0 in the copied {@link ByteBuffer} and the mark if defined is discarded.
      *
      * @param byteBuffer ByteBuffer to send to the service.
      * @return AsyncRequestBody instance.
      */
     static AsyncRequestBody fromByteBuffer(ByteBuffer byteBuffer) {
-        return fromBytes(BinaryUtils.copyAllBytesFrom(byteBuffer));
+        ByteBuffer immutableCopy = BinaryUtils.immutableCopyOf(byteBuffer);
+        immutableCopy.rewind();
+        return ByteBuffersAsyncRequestBody.of(null, immutableCopy);
+    }
+
+    /**
+     * Creates an {@link AsyncRequestBody} from a {@link ByteBuffer} <b>without</b> copying the contents of the
+     * {@link ByteBuffer}. This introduces concurrency risks, allowing: (1) the caller to modify the {@link ByteBuffer} stored in
+     * this {@code AsyncRequestBody} implementation AND (2) any users of {@link #fromByteBufferUnsafe(ByteBuffer)} to modify the
+     * {@link ByteBuffer} passed into this {@code AsyncRequestBody} implementation.
+     *
+     * <p>As the method name implies, this is unsafe. Use {@link #fromByteBuffer(ByteBuffer)}} unless you're sure you know the
+     * risks.
+     *
+     * @param byteBuffer ByteBuffer to send to the service.
+     * @return AsyncRequestBody instance.
+     */
+    static AsyncRequestBody fromByteBufferUnsafe(ByteBuffer byteBuffer) {
+        return ByteBuffersAsyncRequestBody.of(null, byteBuffer);
+    }
```
Sorry for the additional comments, since it felt like we were done, but I've just realized that semantics of these two seem to be different for the position. It seems like fromByteBuffer starts from position 0 (for some reason) and fromByteBufferUnsafe does not (which seems more correct).

Should we do the (probably less correct) thing for `fromByteBufferUnsafe` with `byteBuffer.asReadOnlyBuffer().rewind()`? (same with `fromByteBuffersUnsafe`)

Later on, we should probably deprecate `fromByteBuffer` for a new (more correct) `fromByteBufferPosition` and `fromByteBufferPositionUnsafe` that actually honors the current position of the buffer.

StephenFlavin replied:
Yes I noticed this too, looking at [`RequestBody`](https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/sync/RequestBody.java#L174-L198) I think adding the implementation for `fromRemainingByteBuffer`, `fromRemainingByteBufferUnsafe` and `fromRemainingByteBufferUnsafe` makes sense.

StephenFlavin replied:
I think this method ensuring it passes along a read only buffer makes sense too to mitigate
>(2) any users of {@link #fromByteBufferUnsafe(ByteBuffer)} to modify the
     * {@link ByteBuffer} passed into this {@code AsyncRequestBody} implementation.

StephenFlavin replied:
I've made the changes and kept them in a [separate commit](https://github.com/aws/aws-sdk-java-v2/pull/3925/commits/1fd1420bfd97d796d2c90f846e61ff7c6e87be0a) to make it easier to review, if you're happy I'll merge these commits.

millems replied:
LGTM, let's merge!

StephenFlavin replied:
commits merged 🙂

millems commented:
Thanks for your contribution! I'm going to update it from the target branch and then run the tests.

millems commented:
`ChecksumCalculatingAsyncRequestBodyTest.bytesBufferConstructorHasCorrectContentType` in `sdk-core` seems to be failing. Can you try to repro on your end?

StephenFlavin commented:
@millems the tests were expecting the body to have the `contentLength` set which by default was `Optional.empty()`
I've changed the static `ByteBuffersAsyncRequestBody#of` method to sum the total `ByteBuffer#remaining` which fixes it and added unit tests to validate it.
Apologies I've not been able to get a full build to complete locally, running into issues with [flakey tests](https://github.com/aws/aws-sdk-java-v2/issues/3926) and python.
I've run the sdk-core and utils units successfully at least once.

millems commented:
@StephenFlavin Awesome, let me kick off the suites again!

StephenFlavin commented:
It looks like some of the builds have failed, is there a way for me to see the output @millems?
I'm assuming this is because of the flakey tests I mentioned

millems commented:
@StephenFlavin Sorry, I was off yesterday. I wish there was a way for you to see the output. We're still iterating on the security of our testing solution, trying to figure out how to expose the logs.

It looks like test failures in `:protocol-tests`, and `:codegen-generated-classes-test` this time. Those modules do functional tests of the SDK clients, so they sometimes catch issues that appear in integration of components. Are you able to reproduce those errors yourself by building the modules? Otherwise I can pull the specific test failures.

StephenFlavin commented:
there does appear to just be some unstable tests on master e.g. https://github.com/aws/aws-sdk-java-v2/commit/c015a19518941d36f3e85ddfc96714d2d95e66c3

StephenFlavin commented:
I've been able to get them running, `:protocol-tests` the only failing tests I have are integration tests because I don't have IAM credentials I can use re: https://github.com/aws/aws-sdk-java-v2/blob/master/docs/GettingStarted.md#integration-tests
The `:codegen-generated-classes-test` tests do seem to have genuine failures in [`AsyncStreamingCoreMetricsTest`](https://github.com/aws/aws-sdk-java-v2/blob/master/test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/metrics/async/AsyncStreamingCoreMetricsTest.java), I'll have a closer look at this one.
```
Request content was only 0 bytes, but the specified content-length was 10 bytes.

StephenFlavin commented:
I've not had a chance to spend time on this, will try to get it done this week

StephenFlavin commented:
@millems test failures fixed.

The issues was `ByteBuffersAsyncRequestBody` didn't support multiple subscribers properly as it was just passing along the same `ByteBuffer` which it was holding a reference to, so if one subscriber read the buffer the other subscriber would attempt to read the same buffer but it had already been consumed, the `Subscriber#onNext` call now does `buffer.asReadOnlyBuffer()` and I've added a test to verify this requirement in future.

This is a change in comparison to how `ByteArrayAsyncRequestBody` worked as it would have been passing along mutable copies but I don't think this is an issue.

To support this I had to fix an issue in `ChunkBuffer.java` where it was expecting `ByteBuffer#array` to work which is an optional operation for `ByteBuffer` implementations.
While fixing this I've identified an issue with `BinaryUtils#copyAllBytesFrom` where it treats implementations that support `ByteBuffer#array` as `BinaryUtils#copyRemainingBytesFrom` because it respects the current position doing 
```java
        if (bb.hasArray()) {
            return Arrays.copyOfRange(
                    bb.array(),
                    bb.arrayOffset(),
                    bb.arrayOffset() + bb.limit());
        }
```
this should just be doing 
```java
        if (bb.hasArray()) {
            return bb.array().clone;
        }
```
I've opted to leave this alone as part of this PR, shall I raise a separate PR for this or shall I open an issue?

millems commented:
Re-running tests.

millems commented:
Thank you for your contribution! It has been merged and will go out with the next regular release.

StephenFlavin commented:
@millems or @debora-ito,
This PR resolves https://github.com/aws/aws-sdk-java-v2/issues/1735 so that can be closed 🙂 
Also would it be possible to be added to the all-contributors thing 🙏

millems commented:
Sorry, hadn't realized we didn't add you!

StephenFlavin commented:
np, thanks 🙂
